### PR TITLE
add lief 0.10.1

### PR DESCRIPTION
**Pull request by intelligide**
Specify library name and version:  **lief/0.10.1**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

Based on https://github.com/bincrafters/conan-lief

### CODE REVIEW

File: recipes/lief/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        "with_macho": True,
+        "with_oat": True,
+        "with_pe": True,
+        "with_vdex": True,
+    }
+    _source_subfolder = "source_subfolder"
+    _build_subfolder = "build_subfolder"
+
+    requires = (
+        "rang/3.1.0@rang/stable",
```
Comment (uilianries) [inline recipes/lief/all/conanfile.py:48]: External packages are not allowed. We need rang in CCI first.

**Comment by SSE4** [comment]
there are many unresolved symbols. maybe `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS` can help?

**Comment by SSE4** [comment]
@intelligide the compiler crashes (ICE) - I guess it's better to just include it raising `ConanInvalidConfiguration`

**Comment by intelligide** [comment]
The ICE happens with `compiler.version=14` and `lief:shared=True` only.

Another user got the same error in 2018 lief-project/LIEF#152 and this PR lief-project/LIEF#153 fixed the problem apparently. But I don't understand how this solves the problem.

**Comment by uilianries** [comment]
@intelligide Have your checked with the author? Maybe it's a known limitation.

File: recipes/lief/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        cmake.build()
+
+    def package(self):
+        self.copy(pattern="LICENSE", dst="licenses", src=self._source_subfolder)
+        cmake = self._configure_cmake()
+        cmake.install()
+        tools.rmdir(os.path.join(self.package_folder, "share"))
+
+    def package_info(self):
+        self.cpp_info.names["pkg_config"] = "LIEF"
```
Comment (uilianries) [inline recipes/lief/all/conanfile.py:111]: usually pkgconfig is lower case, do you have any information about upper case for LIEF ?

File: recipes/lief/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        self.cpp_info.names["cmake_find_package"] = "LIEF"
+        self.cpp_info.names["cmake_find_package_multi"] = "LIEF"
+        self.cpp_info.libs = tools.collect_libs(self)
+        self.cpp_info.defines = ["_GLIBCXX_USE_CXX11_ABI=1"]
```
Comment (uilianries) [inline recipes/lief/all/conanfile.py:115]: First time I _GLIBCXX_USE_CXX11_ABI exported, why? I think it should be managed by the consumer.

Comment (intelligide) [reply recipes/lief/all/conanfile.py:115]: It is publicly exported here https://github.com/lief-project/LIEF/blob/master/CMakeLists.txt#L345. Without this definition, LIEF can't be linked on linux.

File: recipes/lief/all/conanfile.py (hunk 2 above)
Comment (intelligide) [reply recipes/lief/all/conanfile.py:111]: LIEF gives information only for cmake (findLIEF). There is no informations for pkgconfig.

**Comment by intelligide** [comment]
> @intelligide Have your checked with the author? Maybe it's a known limitation.

the authors know the problem as you can see here https://github.com/lief-project/LIEF/issues/152#issuecomment-376741942 they don't provide a fix. VS 2015 isn't officially supported.

File: recipes/lief/all/conanfile.py (hunk 3 above)
Comment (uilianries) [reply recipes/lief/all/conanfile.py:115]: thanks for clarifying

**Review by SSE4 — APPROVED** [review]

**Review by danimtb — APPROVED** [review]

File: recipes/lief/all/conanfile.py (hunk 2 above)
Comment (SpaceIm) [reply recipes/lief/all/conanfile.py]: in master, pkg-config file is `LIEF.pc`

File: recipes/lief/all/conanfile.py (hunk 3 above)
Comment (SpaceIm) [reply recipes/lief/all/conanfile.py:115]: Should it raise ConanInvalidConfiguration if self.settings.compiler.libcxx = libstdc++ ?

**Pull request merged**